### PR TITLE
Updated bazel rules tests and added new test cases

### LINKED ISSUE #66: Update bazel rules
Update versions and checksums for bazel rules like jvm_rules_external, rules_scala, rules_kotlin, etc.

### PR DESCRIPTION
**Pull request by mikkoziel**
Fixes #66

### CODE REVIEW

File: e2e/src/test/resources/rules/trivial/rules_apple/WORKSPACE.bzlignore [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    "@build_bazel_apple_support//lib:repositories.bzl",
+    "apple_support_dependencies",
+)
+
+apple_support_dependencies()
+
+# maven deps
+load("@rules_jvm_external//:defs.bzl", "maven_install")
+
+maven_install(
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_apple/WORKSPACE.bzlignore:57]: why maven rules with apple rules?

File: e2e/src/test/resources/rules/trivial/rules_cc/WORKSPACE.bzlignore [hunk 2]
```diff
@@ -0,0 +1,45 @@
+workspace(name = "maven-trivial")
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_cc/WORKSPACE.bzlignore:9]: why jvm rules with cc rules?

File: e2e/src/test/resources/rules/trivial/rules_closure/WORKSPACE.bzlignore [hunk 3]
```diff
@@ -0,0 +1,46 @@
+workspace(name = "maven-trivial")
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_closure/WORKSPACE.bzlignore:5]: why

File: e2e/src/test/resources/rules/trivial/rules_docker/WORKSPACE.bzlignore [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
+    strip_prefix = "rules_jvm_external-{}".format(RULES_JVM_EXTERNAL_TAG),
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_docker/WORKSPACE.bzlignore:12]: why

File: e2e/src/test/resources/rules/trivial/rules_dotnet/WORKSPACE.bzlignore [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
+    strip_prefix = "rules_jvm_external-{}".format(RULES_JVM_EXTERNAL_TAG),
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_dotnet/WORKSPACE.bzlignore:12]: why

File: e2e/src/test/resources/rules/trivial/rules_foreign_cc/WORKSPACE.bzlignore [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_foreign_cc/WORKSPACE.bzlignore:11]: why

File: e2e/src/test/resources/rules/trivial/rules_go/WORKSPACE.bzlignore [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_go/WORKSPACE.bzlignore:11]: why

File: e2e/src/test/resources/rules/trivial/rules_k8s/WORKSPACE.bzlignore [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
+    strip_prefix = "rules_jvm_external-{}".format(RULES_JVM_EXTERNAL_TAG),
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_k8s/WORKSPACE.bzlignore:12]: why

File: e2e/src/test/resources/rules/trivial/rules_proto/WORKSPACE.bzlignore [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
+    strip_prefix = "rules_jvm_external-{}".format(RULES_JVM_EXTERNAL_TAG),
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/rules_proto/WORKSPACE.bzlignore:12]: why?

File: kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt [hunk 10]
```diff
@@ -32,6 +32,7 @@ class BazelRulesExtractor(private val workspaceRoot: Path) {
     val name: String?,
     val sha256: String?,
     val url: String?,
+    val urls: List<String?>?,
```
Comment (lukaszwawrzyk) [inline kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt:35]: List of nulls is possible?

File: kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt [hunk 11]
```diff
@@ -66,9 +67,20 @@ class BazelRulesExtractor(private val workspaceRoot: Path) {
...
+            it.generator_function.isEmpty() &&
+            (!it.url.isNullOrEmpty() || !it.urls.isNullOrEmpty()) &&
+            !it.sha256.isNullOrEmpty()
+        }
+        .mapNotNull {
+          if (!it.url.isNullOrEmpty()) {
+            RuleLibraryId.from(it.url, it.sha256!!)
+          } else {
+            it.urls!!
+              .first { it1 -> it1!!.startsWith("https://github.com/") }
```
Comment (lukaszwawrzyk) [inline kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt:80]: !! should not be needed after fixing my other comment.
Also there should be better names than it and it1

File: e2e/src/test/resources/rules/trivial/rules_apple/WORKSPACE.bzlignore (hunk 1 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_apple/WORKSPACE.bzlignore:57]: Removed

File: e2e/src/test/resources/rules/trivial/rules_cc/WORKSPACE.bzlignore (hunk 2 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_cc/WORKSPACE.bzlignore:9]: Removed

File: e2e/src/test/resources/rules/trivial/rules_closure/WORKSPACE.bzlignore (hunk 3 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_closure/WORKSPACE.bzlignore:5]: Removed

File: e2e/src/test/resources/rules/trivial/rules_dotnet/WORKSPACE.bzlignore (hunk 5 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_dotnet/WORKSPACE.bzlignore:12]: Removed

File: e2e/src/test/resources/rules/trivial/rules_foreign_cc/WORKSPACE.bzlignore (hunk 6 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_foreign_cc/WORKSPACE.bzlignore:11]: Removed

File: e2e/src/test/resources/rules/trivial/rules_go/WORKSPACE.bzlignore (hunk 7 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_go/WORKSPACE.bzlignore:11]: Removed

File: e2e/src/test/resources/rules/trivial/rules_java/WORKSPACE.bzlignore [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
+    strip_prefix = "rules_jvm_external-{}".format(RULES_JVM_EXTERNAL_TAG),
```
[the remark this answers is not in the document]
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_java/WORKSPACE.bzlignore:12]: Removed

File: e2e/src/test/resources/rules/trivial/rules_k8s/WORKSPACE.bzlignore (hunk 8 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_k8s/WORKSPACE.bzlignore:12]: Removed

File: e2e/src/test/resources/rules/trivial/rules_proto/WORKSPACE.bzlignore (hunk 9 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_proto/WORKSPACE.bzlignore:12]: Removed

File: e2e/src/test/resources/rules/trivial/rules_python/WORKSPACE.bzlignore [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
+RULES_JVM_EXTERNAL_SHA = "735602f50813eb2ea93ca3f5e43b1959bd80b213b836a07a62a29d757670b77b"
+
+http_archive(
+    name = "rules_jvm_external",
+    sha256 = RULES_JVM_EXTERNAL_SHA,
```
[the remark this answers is not in the document]
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_python/WORKSPACE.bzlignore:11]: Removed

File: e2e/src/test/resources/rules/trivial/rules_skylib/WORKSPACE.bzlignore [hunk 14]
```diff
@@ -0,0 +1,61 @@
+workspace(name = "maven-trivial")
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+# rules_jvm_external - for maven dependencies
+RULES_JVM_EXTERNAL_TAG = "4.4.2"
```
[the remark this answers is not in the document]
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/rules_skylib/WORKSPACE.bzlignore:6]: Removed

File: kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt (hunk 10 above)
Comment (mikkoziel) [reply kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt:35]: FIxed

File: kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt (hunk 11 above)
Comment (mikkoziel) [reply kinds/bazel/rules/src/main/kotlin/org/virtuslab/bazelsteward/bazel/rules/BazelRulesExtractor.kt:80]: Fixed

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt [hunk 15]
```diff
@@ -29,7 +30,7 @@ open class E2EBase {
   protected fun branch(libraryId: String, version: String): String =
     "$branchRef/$libraryId/$version"
 
-  protected fun expectedBranches(vararg libs: Pair<String, String>): List<String> {
```
Comment (lukaszwawrzyk) [inline e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt:32]: why public?

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt [hunk 16]
```diff
@@ -45,7 +46,7 @@ open class E2EBase {
     runBazelStewardWith(workspaceRoot, args) { x -> x }
   }
 
-  protected fun runBazelStewardWith(
```
Comment (lukaszwawrzyk) [inline e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt:48]: why public?

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+  private val expectedVersions: Array<Pair<String, String>>
+) : E2EBase() {
+
+  @Test
+  fun `Project with specific rules`(@TempDir tempDir: Path) {
+    runBazelStewardWith(tempDir, project) {
+      it.withRulesOnly()
+    }
+
+    val expectedBranches = E2EBase().expectedBranches(*expectedVersions)
```
Comment (lukaszwawrzyk) [inline e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt:18]: You are inheriting from E2EBase, you can just invoke this method directly.

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt [hunk 18]
```diff
@@ -0,0 +1,22 @@
+package org.virtuslab.bazelsteward.e2e
+
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.io.TempDir
+import java.nio.file.Path
+
+open class RulesUpdate(
+  private val project: String,
+  private val expectedVersions: Array<Pair<String, String>>
```
Comment (lukaszwawrzyk) [inline e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt:9]: `private vararg val expectedVersions: Pair<String, String>`

File: e2e/src/test/resources/rules/trivial/all/WORKSPACE.bzlignore [hunk 19]
```diff
@@ -0,0 +1,293 @@
+workspace(name = "maven-trivial")
+
```
Comment (lukaszwawrzyk) [inline e2e/src/test/resources/rules/trivial/all/WORKSPACE.bzlignore:2]: is this used?

**Review by lukaszwawrzyk — CHANGES_REQUESTED** [review]

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt (hunk 15 above)
Comment (mikkoziel) [reply e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt:32]: Changed to protected

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt (hunk 16 above)
Comment (mikkoziel) [reply e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/E2EBase.kt:48]: Changed to protected

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt (hunk 17 above)
Comment (mikkoziel) [reply e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt:18]: Fixed

File: e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt (hunk 18 above)
Comment (mikkoziel) [reply e2e/src/test/kotlin/org/virtuslab/bazelsteward/e2e/RulesUpdate.kt:9]: FIxed

File: e2e/src/test/resources/rules/trivial/all/WORKSPACE.bzlignore (hunk 19 above)
Comment (mikkoziel) [reply e2e/src/test/resources/rules/trivial/all/WORKSPACE.bzlignore:2]: Removed

Comment (lukaszwawrzyk) [reply e2e/src/test/resources/rules/trivial/all/WORKSPACE.bzlignore]: what is removed?

**Review by lukaszwawrzyk — APPROVED** [review]

**Pull request merged**